### DISCUSSION TITLE
Can my IOT devices using CoAPS over OpenThread act as both Client and Server?

### DISCUSSION THREAD

**Discussion by michaelsimp** [openthread/openthread#10321] [Q&A]
A few years back I developed a distributed control system using CoAP over an OpenThread Mesh network for LAN coms on Silicon Labs EFR32 MCUs constrained REEDs. 
An OpenThread Border Router app running on a PC (or SBC) enabled an app on the same PC to communicate to my Thread network.
Application wise my PC app needed bidirectional coms with each REED, but the REEDs did not communicate with each other.
The PC application and the REEDs both have end points eg
- The REEDs send status updates and transactions to the PC application end points
- The PC sends requests and commands to the REED End points.

Now I need to remove my PC application from the site and host it in the cloud. My border router will remain on site but migrate to a more humbler platform eg ESP32 and connect my Thread Mesh network to a Broadband Router or a Cell 4G LTE modem.  I will need to change to CoAPS
I am not sure if my architecture is feasible as my cloud application needs to be able hit my REED endpoints.
MQTT by comparison uses TCP and so after the Nodes establishes a TCP socket, it can be used for unsolicited communications in both directions.
Is it possible to have similar functionality using CoAP network combining both client and server functionality and if so, how might I go about it. eg
- Is it possible to configure my OTBR to automatically convert CoAP on UDP to TCP so I can maintain a TCP socket?
- Is a VPN a viable option for facilitating client to server communications in both directions?
- I guess another option is for the REEDs to poll my cloud application server regularly to see if they have any requests to make but I am not sure if this is a bit of a hack?

If there is a viable solution, can you point me to any documentation covering this, or give me a steer in the right direction as to how I should approach this.

Thanks

**Comment by jwhui** [comment]
> Is it possible to configure my OTBR to automatically convert CoAP on UDP to TCP so I can maintain a TCP socket?

OpenThread does not currently implement CoAP over TCP. The most common use of CoAP I've seen is with UDP.

> Is a VPN a viable option for facilitating client to server communications in both directions?

A VPN is possible. VPN allows the Thread network to use a Unique Local Prefix rather than relying on global prefixes.

> I guess another option is for the REEDs to poll my cloud application server regularly to see if they have any requests to make but I am not sure if this is a bit of a hack?

Having REEDs poll the cloud application is more inline with the use case CoAP was designed for - which is to support RESTful APIs.
